### PR TITLE
Refactor query engine tool code and use auto_routed mode for LlamaCloudIndex

### PR DESCRIPTION
**Pull request by leehuwuj**
Fix: https://github.com/run-llama/create-llama/issues/434

## Summary by CodeRabbit

- **New Features**
	- Introduced a streamlined query engine tool creation process across various components.
	- Added support for a single query engine tool instead of multiple tools in workflows.
	- Enhanced functionality for the LlamaCloudIndex with auto-routing capabilities.

- **Bug Fixes**
	- Improved error handling and parameter management in query engine functions.

- **Documentation**
	- Updated function signatures and descriptions to reflect changes in tool management.

- **Refactor**
	- Simplified the logic in multiple functions to enhance modularity and readability.
	- Removed unnecessary parameters and improved the overall flow of tool creation.

- **Chores**
	- Cleaned up import statements and organized code structure for better maintainability.

### CODE REVIEW

File: templates/components/engines/python/agent/tools/query_engine.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    """
+    query_engine = create_query_engine(index, engine_params)
+    return QueryEngineTool.from_defaults(
+        query_engine=query_engine,
+        name=tool_name,
+        description=tool_description,
+    )
+
+
+def get_tools(**kwargs):
```
Comment (marcusschiesser) [inline templates/components/engines/python/agent/tools/query_engine.py:42]: Where is this used?

File: templates/types/streaming/express/src/controllers/chat.controller.ts [hunk 2]
```diff
@@ -24,8 +25,12 @@ export const chat = async (req: Request, res: Response) => {
 
     // retrieve document ids from the annotations of all messages (if any)
     const ids = retrieveDocumentIds(messages);
-    // create chat engine with index using the document ids
-    const chatEngine = await createChatEngine(ids, data);
+    // create chat engine with index using the document ids and additional params passed in the request body
```
Comment (marcusschiesser) [inline templates/types/streaming/express/src/controllers/chat.controller.ts:28]: The disadvantage here is that it complicates the example code

File: templates/components/multiagent/python/app/api/routers/chat.py [hunk 3]
```diff
@@ -25,10 +26,13 @@ async def chat(
 
         doc_ids = data.get_chat_document_ids()
         filters = generate_filters(doc_ids)
-        params = data.data or {}
+        params = data.data or {
```
Comment (marcusschiesser) [inline templates/components/multiagent/python/app/api/routers/chat.py:29]: Data could overwrite the filters?

File: templates/components/engines/typescript/agent/tools/query-engine.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+      name: "query_engine",
+      description: `Use this tool to retrieve information about the text corpus from an index.`,
+      ...metadata,
+    },
+  });
+}
+
+function createQueryEngine(
+  index: VectorStoreIndex | LlamaCloudIndex,
+  params: any = {},
```
Comment (marcusschiesser) [inline templates/components/engines/typescript/agent/tools/query-engine.ts:25]: Doesn't LITS have a better type there?

File: templates/components/engines/typescript/agent/tools/query-engine.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  BaseQueryEngine,
+  LlamaCloudIndex,
+  QueryEngineTool,
+  VectorStoreIndex,
+} from "llamaindex";
+
+export function createQueryEngineTool(
+  index: VectorStoreIndex | LlamaCloudIndex,
+  params?: any,
+  metadata?: any,
```
Comment (marcusschiesser) [inline templates/components/engines/typescript/agent/tools/query-engine.ts:11]: Stricter types?

File: templates/components/engines/python/agent/tools/query_engine.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    # use auto_routed mode for better query results
+    if index.__class__.__name__ == "LlamaCloudIndex":
+        params["auto_routed"] = True
+    return index.as_query_engine(**params)
+
+
+def get_query_engine_tool(
+    index,
+    engine_params: Optional[Dict[str, Any]] = None,
+    tool_name: Optional[str] = None,
```
Comment (marcusschiesser) [inline templates/components/engines/python/agent/tools/query_engine.py:22]: As we create a tool, how about removing the tool_ prefix?

File: templates/components/engines/python/agent/tools/query_engine.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+def create_query_engine(index, params: Optional[Dict[str, Any]] = None):
+    params = params or {}
+    top_k = int(os.getenv("TOP_K", 0))
+    if top_k != 0:
+        params["similarity_top_k"] = top_k
+    # If index is index is LlamaCloudIndex
+    # use auto_routed mode for better query results
+    if index.__class__.__name__ == "LlamaCloudIndex":
+        params["auto_routed"] = True
```
Comment (marcusschiesser) [inline templates/components/engines/python/agent/tools/query_engine.py:15]: What's actual the default?

File: templates/components/engines/python/agent/engine.py [hunk 8]
```diff
@@ -1,28 +1,27 @@
...
     callback_manager = CallbackManager(handlers=event_handlers or [])
 
     # Add query tool if index exists
     index_config = IndexConfig(callback_manager=callback_manager, **(params or {}))
     index = get_index(index_config)
     if index is not None:
-        query_engine = index.as_query_engine(
-            filters=filters, **({"similarity_top_k": top_k} if top_k != 0 else {})
-        )
+        query_engine = create_query_engine(index, params)
```
Comment (marcusschiesser) [inline templates/components/engines/python/agent/engine.py:24]: Why not using the factory for the tool?

File: templates/components/agents/python/form_filling/app/workflows/form_filling.py [hunk 9]
```diff
@@ -23,25 +14,37 @@
...
+    chat_with_tools,
+)
 
-def create_workflow(
-    chat_history: Optional[List[ChatMessage]] = None,
-    params: Optional[Dict[str, Any]] = None,
-    filters: Optional[List[Any]] = None,
-) -> Workflow:
+
+def _create_query_engine_tool(params=None) -> QueryEngineTool:
```
Comment (marcusschiesser) [inline templates/components/agents/python/form_filling/app/workflows/form_filling.py:27]: Why not using the tool factory?

File: templates/components/engines/python/agent/tools/query_engine.py (hunk 7 above)
Comment (leehuwuj) [reply templates/components/engines/python/agent/tools/query_engine.py:15]: The default is None, which is not mentioned in the client code, but it seems to me that it'll fall back to `chunk` mode.

File: templates/components/multiagent/python/app/api/routers/chat.py (hunk 3 above)
Comment (leehuwuj) [reply templates/components/multiagent/python/app/api/routers/chat.py:29]: great catch!

File: templates/components/engines/python/agent/engine.py (hunk 8 above)
Comment (leehuwuj) [reply templates/components/engines/python/agent/engine.py:24]: yes, it's a TBD part.  
Generally, we can initialize the index from the `tools/query_engine.py` code and use tool factory for query engine like other tools. But the tools factory by default is creating the tool from the config file, so I think better to inject the index from the run time to create query engine tool instead.

File: templates/types/streaming/express/src/controllers/chat.controller.ts (hunk 2 above)
Comment (leehuwuj) [reply templates/types/streaming/express/src/controllers/chat.controller.ts]: Yes, after checking the type, I kept this file as is, which is easier for the user to understand.

File: templates/components/agents/python/blog/app/agents/researcher.py [hunk 10]
```diff
@@ -1,39 +1,28 @@
...
-from llama_index.core.tools import QueryEngineTool, ToolMetadata
+from llama_index.core.tools import QueryEngineTool
+from app.engine.tools.query_engine import get_query_engine_tool
 
 
-def _create_query_engine_tool(params=None) -> QueryEngineTool:
-    """
-    Provide an agent worker that can be used to query the index.
-    """
+def _create_query_engine_tool(
```
Comment (marcusschiesser) [inline templates/components/agents/python/blog/app/agents/researcher.py:12]: isn't that simpler to move this code to `_get_research_tools`?

File: templates/components/agents/python/financial_report/app/workflows/financial_report.py [hunk 11]
```diff
@@ -24,19 +23,23 @@
 )
 
 
+def _create_query_engine_tool(params=None, **kwargs) -> QueryEngineTool:
```
Comment (marcusschiesser) [inline templates/components/agents/python/financial_report/app/workflows/financial_report.py:26]: isn't that simpler to move this code to `create_workflow`?

File: templates/components/agents/python/form_filling/app/workflows/form_filling.py [hunk 12]
```diff
@@ -23,25 +14,35 @@
...
+    chat_with_tools,
+)
+
 
-def create_workflow(
-    chat_history: Optional[List[ChatMessage]] = None,
-    params: Optional[Dict[str, Any]] = None,
-    filters: Optional[List[Any]] = None,
-) -> Workflow:
+def _create_query_engine_tool(
```
Comment (marcusschiesser) [inline templates/components/agents/python/form_filling/app/workflows/form_filling.py:27]: isn't that simpler to move this code to `create_workflow`?

**Pull request merged**